server/src/main/java/org/elasticsearch/rest/AbstractRestChannel.java

```diff
@@ -137,25 +141,30 @@ public XContentBuilder newBuilder(@Nullable XContentType requestContentType, @Nu
 
     /**
      * A channel level bytes output that can be reused. The bytes output is lazily instantiated
-     * by a call to {@link #newBytesOutput()}. Once the stream is created, it gets reset on each
-     * call to this method.
+     * by a call to {@link #newBytesOutput()}. This method should only be called once per request.
      */
     @Override
     public final BytesStreamOutput bytesOutput() {
-        if (bytesOut == null) {
-            bytesOut = newBytesOutput();
-        } else {
-            bytesOut.reset();
+        if (bytesOut != null) {
+            // fallback in case of encountering a bug, release the existing buffer if any (to avoid leaking memory) and acquire a new one
+            // to send out an error response
+            assert false : "getting here is always a bug";
+            logger.error("channel handling [{}] reused", request.rawPath());
+            releaseOutputBuffer();
```

Comment:
I think we need this in practice to prevent bad things from happening in case of serialisation bugs (you don't even have to get into `sendResponse` to acquire the buffer here). I don't want to throw `AssertionError` here and potentially kill the node, throw anything and thus leak memory or other resources associated with the channel -> this was the next best option I could think of.